### ISSUE TITLE
Make `transform` round-trip emit the same BAM file

### ISSUE THREAD

**Issue by ryan-williams** [bigdatagenomics/adam#794]
When testing #790, I was thinking that it would be nice if we made `transform`ing a BAM through ADAM and back return the same file, e.g.:

```
adam transform foo.bam foo.adam
adam transform foo.adam foo.2.bam
diff -q foo.bam foo.2.bam
```

Attempting this on toy `.sam` files found in the ADAM repo yields different `.sam` files that are non-trivial (and basically impossible programmatically) to verify are the same as their corresponding originals, which is disorienting!

I noticed four changes we make to BAMs today:

1. we drop `@SQ` lines that aren't mapped to in the file
1. I'm not sure that we preserve the same `@SQ` sort order, if they are not in lex order in the original BAM, nor do we support sorting reads according to an arbitrary `@SQ` order (per discussion on #760).
1. optional tags are not output in the same order.
1. we add an `M5` tag to `@SQ` lines that equals the `SN` field, which seems wrong; why add incorrect MD5 sum info?

I think that is roughly in increasing order of difficulty to fix.

This is by no means essential, but maybe worth working towards a bit, under the theme of making it easy/painless as possible for people to get data in and out of ADAM.

**Comment by fnothaft** [comment]
I think this is reasonable for sorted output.

**Comment by ryan-williams** [comment]
By "sorted output" do you mean "cases where the original BAM already has its `@SQ`s lex-sorted?"

**Comment by fnothaft** [comment]
Ish... More so, I mean BAM files with `@HD order:coordinate`.

**Comment by ryan-williams** [comment]
Interesting; any reason that it doesn't make sense to have identical round-trips with e.g. `@HD SO:queryname`?

btw, this happened:

![](http://cl.ly/image/2U250I1T3Z0K/Screen%20Shot%202015-08-21%20at%202.19.00%20PM.png)

probably just a gmail fluke

**Comment by fnothaft** [comment]
Ah, sorry, I just generally meant !`@HD SO:undefined`. `SO:queryname` should be A-OK as well.

I am indeed hacking your gmail via GitHub... They've caught me! ;)

**Comment by ryan-williams** [comment]
Hm, I just noticed a 4th thing we do differently: we add an `M5` tag to `@SQ` lines that equals the `SN` field, which seems wrong; why add incorrect MD5 sum info?

**Comment by fnothaft** [comment]
> Hm, I just noticed a 4th thing we do differently: we add an M5 tag to @SQ lines that equals the SN field, which seems wrong; why add incorrect MD5 sum info?

Oh, wow. That is just a bug... Someone must've meant to add the MD5 sum that we keep in `SequenceRecord` but made an error when doing it...

**Comment by ryan-williams** [comment]
One more: we `0` out the "pos" on unmapped reads; seems like we could just not do that and it wouldn't necessarily harm anything. If people are e.g. denormalizing mapped-mate-start into that field, do we care that much?

**Comment by fnothaft** [comment]
The [SAM spec](https://samtools.github.io/hts-specs/SAMv1.pdf) seems pretty unambiguous to me about 0'ing out POS on unmapped reads. The SPEC is a bit vaguer for CIGARs than for POS, but it still seems to me like it is pretty clear that CIGAR should be set to `*` from the spec.

Perhaps I'm misunderstanding what you're saying, but why would they put mate start in POS? That's what the SAM/BAM PNEXT field is for?

**Comment by ryan-williams** [comment]
> Perhaps I'm misunderstanding what you're saying, but why would they put mate start in POS? That's what the SAM/BAM PNEXT field is for?

Yea, you're right, it doesn't really make sense for someone to want/need to use POS that way. I am just trying to make sense of e.g. [bqsr1.sam](https://github.com/bigdatagenomics/adam/blob/633fdf9d0073417e6f2f42ea54ee29447a9f3dd2/adam-cli/src/test/resources/bqsr1.sam), which:
* has no `SO` sort-order defined,
* has 8 unmapped reads with all mapped fields mimicking those of their mapped mates,
* given that, obeys "coordinate" sort: the unmapped reads are next to their mapped mates,
* does not obey "queryname" sort order.

> The SAM spec seems pretty unambiguous to me about 0'ing out POS on unmapped reads. The SPEC is a bit vaguer for CIGARs than for POS, but it still seems to me like it is pretty clear that CIGAR should be set to * from the spec.

Yea, it sounds like we just have arguably-malformed sam's in the ADAM repo that I've been using as test cases.

1. Maybe we should fix that, and then we can be more consistently opinionated about some of these assumptions, and get properties like round-trip `transform`ing for free.
2. I assume lots of tools output malformed SAMs like this; their existence here in the first place seems to imply that, at the very least.

**Comment by ryan-williams** [comment]
The spec's use of "coordinate" is pretty confusing.

> POS: 1-based leftmost mapping POSition of the first matching base. The first base in a reference sequence has coordinate 1. POS is set as 0 for an unmapped read without coordinate. If POS is 0, no assumptions can be made about RNAME and CIGAR.

What could "POS is set as 0 for an unmapped read without coordinate" mean?

**Comment by fnothaft** [comment]
> The spec's use of "coordinate" is pretty confusing.
> What could "POS is set as 0 for an unmapped read without coordinate" mean?

I think "without coordinate" is supposed to read as "without a reference name". For me, I would read this as translating "an unmapped read without coordinate" to "a read not mapped to a reference contig". 

> 1. Maybe we should fix that, and then we can be more consistently opinionated about some of these assumptions, and get properties like round-trip transforming for free.

That sounds reasonable to me. OOC, do you have the read name for one of the malformed reads in `bqsr1.sam`? I'd be interested in doing some of this cleanup.

> 2. I assume lots of tools output malformed SAMs like this; their existence here in the first place seems to imply that, at the very least.

I think that most files that are _really_ malformed throw exceptions from inside of htsjdk on read. For files that are in non-compliance with the SAM spec, but that aren't unreadable, I think it is reasonable to say that we can import them, but that we can't make any round trip guarantees.

As an aside, @erictu and I found a pretty nasty bug recently though where htsjdk wouldn't read a file (I think it was a vcf.gz or a bcf?) that was generated from htslib.

**Comment by ryan-williams** [comment]
> I think "without coordinate" is supposed to read as "without a reference name".

OK, I buy that, but that means that the 8 unmapped reads in bqsr1.sam are "malformed".

> OOC, do you have the read name for one of the malformed reads in bqsr1.sam? I'd be interested in doing some of this cleanup.

[Here's one](https://github.com/bigdatagenomics/adam/blob/633fdf9d0073417e6f2f42ea54ee29447a9f3dd2/adam-cli/src/test/resources/bqsr1.sam#L347), I posted it a couple comments back but there's lots of info flying around :)

Here's all of them:

```
$ samtools view -f 4 adam-cli/src/test/resources/bqsr1.sam
SRR062634.20911784	133	22	16060584	0	35M65S	=	16060584	0	TGTAGTGGCAGGGGCCCGTTATCCCAAACTACCTGGGGGGGGGGGGGGGGGGGAACACCTAAAACCCGGGGGGGGGGGGGTTGGTGGGGGCTTTATCGC	GGGGGGGDGG@#########################################################################################	RG:Z:SRR062634	XC:i:35
SRR062634.9119161	133	22	16062369	0	*	=	16062369	0	TTATATGTGTTTTTAAACTAAACTAATTTTATAGGAAAAATAATTTCTTTCCTTCCCTGTTATATCAAATACAGCCTTTAGCTCAAGACACAAGTAATT	CCAC?BD?DCCD:CCACD5=DCDAB=AA=5DBC=BC:??BB85:CBBB<7<BBBBBB?A?BBB5BB:BBBB?B?BBB@??:=0====::'>;'25?????	RG:Z:SRR062634
SRR062634.4789722	133	22	16071485	0	35M65S	=	16071485	0	ATAATATAATGAAAAATAGAGACTGAGAGAGAAAAAACAAGACCCTTTATCTTTATATTTTTTCATATGTGTTTTTTTTTCTGTCTGGTTTTTTTGTTT	####################################################################################################	RG:Z:SRR062634	XC:i:35
SRR062634.18958430	133	22	16076921	0	35M65S	=	16076921	0	TTTTTTCCTGTCTTGGTTGTATAAAAAAAGAGGGAGAAACGCCTGGCAGGGCACCCCAACAAAGGAAGGGAGGAGGGGGTCCCAAGGGGGCCCCGCGGG	####################################################################################################	RG:Z:SRR062634	XC:i:35
SRR062634.10448889	117	22	16079761	0	*	=	16079761	0	TTTCTTTCTTTTATATATATATACACACACACACACACACACACACATATATGTATATATACACGTATATGTATGTATATATGTATATATACACGTATA	@DF>C;FDC=EGEGGEFDGEFDD?DFDEEGFGFGGGDGGGGGGGEGGGGFGGGFGGGGGGFGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG	RG:Z:SRR062634
SRR062634.16769670	117	22	16081625	0	*	=	16081625	0	GGTGAGATGATTGCTGGGATTACAGGCGTGAGCCACCGCGCCTGGCCGTATGTTTATTCTTATGATAGTACCATACTGTTTTGTAGTATGTTTTATAGC	CCCCBCC:CA:CDCBDCBCCA?C@ABBCCBADFECFGGGGGGGGEGDGGGGGGGGFDGGGGGGGGGGGGEGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG	RG:Z:SRR062634
SRR062634.17698657	117	22	16083363	0	*	=	16083363	0	ATAACATTTGGTCTATAGCATCAGAGCCTTATGCACTCAGAGGAAATCCAAAATCACCTATAAGTATTTGCTGGTCCCCTCTGGGCTTAGGGAAATCTC	AAB8AA@A=C>=>>:=:5=;5;*83:=??@7:21+88748:@CBBC:CCC<<>17,>C>>@@C??>>A???A:GGGGGGGEGGGGGGGGGGGGGGGGGGG	RG:Z:SRR062634
SRR062634.17969132	133	22	16096818	0	35M65S	=	16096818	0	TGGGCCCCGTGGACCCTGGCGACCCCCGGGGGAGGCCCGGGGGGGCCCCCTGGCCCCCAAGGGGGGGCCCCAACGGGGAGAAGGGTCCCTAGGGGGGGG	GGGGGGGGGFGGGGGGEEGEAA<@A###########################################################################	RG:Z:SRR062634	XC:i:35
```

> For files that are in non-compliance with the SAM spec, but that aren't unreadable, I think it is reasonable to say that we can import them, but that we can't make any round trip guarantees.

Makes sense. I guess we can agree that `transform` should be idempotent, i.e. applying it twice should give the same thing as applying it once. I've not found any violations of that in the ADAM repo but have only checked a couple `.sam`s.

**Comment by fnothaft** [comment]
> OK, I buy that, but that means that the 8 unmapped reads in bqsr1.sam are "malformed".

Yeah, it's a bit unsatisfying, esp. since the files were generated with BWA, which _should_ be spec compliant:

```
fnothaft$ grep "^@PG" ../adam/adam-core/src/test/resources/bqsr1.sam 
@PG	ID:bwa	PN:bwa	VN:0.7.5a-r405
```

Perhaps I am misreading the spec?

> Here's one, I posted it a couple comments back but there's lots of info flying around :)

Thanks! I'd missed it in the comment upstream...

> Makes sense. I guess we can agree that transform should be idempotent, i.e. applying it twice should give the same thing as applying it once.

Just for import/export, you mean? For that I would definitely agree, but certain things in `transform` (e.g., BQSR and indel realignment) definitely aren't idempotent.

**Comment by ryan-williams** [comment]
> Yeah, it's a bit unsatisfying, esp. since the files were generated with BWA, which should be spec compliant:

Well, looking at the spec more closely, "POS is set as 0 for an unmapped read without coordinate" could just mean that [RNAME == `*`] ⇒ [POS == `0`]. If you set RNAME on an unmapped read, then you can set POS too, according to this interpretation.

> Just for import/export, you mean?

Yea I guess I'm just talking about simple bam⟷adam conversions.

**Comment by fnothaft** [comment]
> If you set RNAME on an unmapped read, then you can set POS too, according to this interpretation.

Yeah, that would seem to line up with these reads, since they have `RNAME == 22`.

> Yea I guess I'm just talking about simple bam⟷adam conversions.

SGTM; this seems like a good goal.

**Comment by ryan-williams** [comment]
So, we've fixed a couple obvious bugs here, and I'm glad we talked this through for posterity.

I think I'm going to redefine the goal here as making ADAM conversion round-trips idempotent *once the data has been ported into ADAM's chosen subset of SAM*.

As far as I know, it does this today (or will reasonably once some of the fixes referenced here are in), so I will close this. If anyone has further things to discuss we can reopen.
